### ISSUE TITLE
Question: How to get API call details inside `Retryer.Retry`?

### ISSUE THREAD

**Issue by erezrokah** [googleapis/gax-go#363]
- What you're trying to do

I'm adding debug logs to our `Retryer` implementation, but I can only log the error and number of retries, not the call itself. That makes it hard to understand which API call is causing retries.

- What code you've already tried

Our retry code looks like this

```go
type Retrier struct {
	backoff    gax.Backoff
	curRetrier int
	maxRetries int
	codes      []codes.Code
	logger     zerolog.Logger
}

func (r *Retrier) Retry(err error) (time.Duration, bool) {
	st, ok := status.FromError(err)
	if !ok {
		r.logger.Debug().Err(err).Msg("failed to get status from error, not retrying")
		return 0, false
	}
	c := st.Code()
	for _, rc := range r.codes {
		if c == rc {
			if r.curRetrier >= r.maxRetries {
				r.logger.Debug().Str("code", c.String()).Int("retry.attempt", r.curRetrier).Err(err).Msg("retried too many times")
				return 0, false
			}
			r.curRetrier++
			r.logger.Debug().Str("code", c.String()).Int("retry.attempt", r.curRetrier).Err(err).Msg("error is retryable, retrying")
			return r.backoff.Pause(), true
		}
	}
	r.logger.Debug().Str("code", c.String()).Err(err).Msg("error is not retryable")
	return 0, false
}
```

- Any error messages you're getting

N/A

Is there a way to get the call details from the error I'm missing, and if not, are you open to passing more information to the retrier?

**Comment by quartzmo** [comment]
Hi @erezrokah,

Thanks for this question.

> Is there a way to get the call details from the error

Depending on your transport (gRPC or HTTP) and the underlying error you're getting, you might find the details you want in the top-level error or a nested error. Take a look at [apierror.go](https://github.com/googleapis/gax-go/blob/main/v2/apierror/apierror.go#L275-L290), starting with `ParseError` (linked), to get an idea of what might be available.

**Comment by erezrokah** [comment]
> Depending on your transport (gRPC or HTTP) and the underlying error you're getting, you might find the details you want in the top-level error or a nested error. Take a look at [apierror.go](https://github.com/googleapis/gax-go/blob/main/v2/apierror/apierror.go#L275-L290), starting with `ParseError` (linked), to get an idea of what might be available.

Thanks @quartzmo, I took at look at what the returned error has and it only has the gRPC status (`FailedPrecondition`) and the error message that doesn't have detailed enough details.
The origin of the error is from Google Cloud, for example https://github.com/googleapis/google-cloud-go/blob/4927b533dd352bee0c3efdc3a88ec96279532e64/accessapproval/apiv1/access_approval_client.go#L681, so is it correct that the server has to add the details to the API error?

Regardless I think this cannot be done in `gax-go` as it only has the `CallSettings` which doesn't say much either.

Looks like using `grpc.WithUnaryInterceptor(logging.UnaryClientInterceptor(...))` is the way.

Closing the issue
